### PR TITLE
add vxlan vni server and client service chain element

### PR DESCRIPTION
**Pull request by pperiyasamy**
This commit moves vxlan vni server and client service chain element from sdk-vpp into sdk repo
so that this can be consumed by other forwarder(s)

Signed-off-by: Periyasamy Palanisamy <periyasamy.palanisamy@est.tech>

## Description

Based on the feedback [here](https://github.com/networkservicemesh/sdk-vpp/issues/215#issuecomment-839806922), moving vni service chain element to sdk repo.

## Issue link

https://github.com/networkservicemesh/sdk-vpp/issues/215

## How Has This Been Tested?

- [ ] Added unit testing to cover
- [ ] Tested manually
- [ ] Tested by integration testing
- [x] Have not tested

## Types of changes

- [ ] Bug fix
- [ ] New functionallity
- [ ] Documentation
- [x] Refactoring
- [ ] CI

### CODE REVIEW

**Comment by edwarnicke** [comment]
@pperiyasamy See [comment](https://github.com/networkservicemesh/sdk-kernel/pull/231#issuecomment-842335241) on how to fix the golanglint-ci errors :)

File: pkg/networkservice/common/mechanisms/vxlan/vni/client.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+// NewClient - set the SrcIP for the vxlan mechanism
+func NewClient(tunnelIP net.IP) networkservice.NetworkServiceClient {
+	return &vniClient{
+		tunnelIP: tunnelIP,
+	}
+}
+
+func (v *vniClient) Request(ctx context.Context, request *networkservice.NetworkServiceRequest, opts ...grpc.CallOption) (*networkservice.Connection, error) {
+	for _, m := range request.GetMechanismPreferences() {
```
Comment (Bolodya1997) [inline pkg/networkservice/common/mechanisms/vxlan/vni/client.go:44]: Basically we expect mechanism client to append its mechanism to `Request.MechanismPreferences`.
Such thing with a loop over the existing preferences is used to guard client if someone have already set needed mechanism but haven't fully filled parameters:
https://github.com/networkservicemesh/sdk/blob/d09236b534179e947c06fabc388a96cae03dc9fe/pkg/networkservice/common/mechanisms/kernel/client.go#L48-L60
1. Try to update
2. If there is nothing to update, then add
So the result of this Request should be needed mechanism in preferences with all needed parameters filled.

File: pkg/networkservice/common/mechanisms/vxlan/vni/server.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	if mechanism := vxlan.ToMechanism(request.GetConnection().GetMechanism()); mechanism != nil {
+		mechanism.SetDstIP(v.tunnelIP)
+		k := vniKey{
+			srcIPString: mechanism.SrcIP().String(),
+			vni:         mechanism.VNI(),
+		}
+		// If we already have a VNI, make sure we remember it, and go on
+		if k.vni != 0 && mechanism.SrcIP() != nil {
+			v.Map.LoadOrStore(k, &k)
+			return next.Server(ctx).Request(ctx, request)
```
Comment (Bolodya1997) [inline pkg/networkservice/common/mechanisms/vxlan/vni/server.go:53]: Probably we need to add here check for returned error and `v.Map.Delete()` if so.

File: pkg/networkservice/common/mechanisms/vxlan/vni/server.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+				return nil, err
+			}
+			// If its not one already in use, set it and we are good to go
+			if _, ok := v.Map.LoadOrStore(k, &k); !ok {
+				mechanism.SetVNI(k.vni)
+				break
+			}
+		}
+	}
+	return next.Server(ctx).Request(ctx, request)
```
Comment (Bolodya1997) [inline pkg/networkservice/common/mechanisms/vxlan/vni/server.go:70]: Same here.

File: pkg/networkservice/common/mechanisms/vxlan/vni/client.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
+// Copyright (c) 2020-2021 Cisco and/or its affiliates.
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/mechanisms/vxlan/vni/client.go:1]:
```suggestion
// Copyright (c) 2021 ${TODO: Add your comany name} and/or its affiliates.
```

File: pkg/networkservice/common/mechanisms/vxlan/vni/doc.go [hunk 5]
```diff
@@ -0,0 +1,18 @@
+// Copyright (c) 2020-2021 Cisco and/or its affiliates.
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/mechanisms/vxlan/vni/doc.go:1]:
```suggestion
// Copyright (c) 2021 ${TODO: Add your comany name} and/or its affiliates.
```

File: pkg/networkservice/common/mechanisms/vxlan/vni/gen.go [hunk 6]
```diff
@@ -0,0 +1,31 @@
+// Copyright (c) 2020-2021 Cisco and/or its affiliates.
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/mechanisms/vxlan/vni/gen.go:1]:
```suggestion
// Copyright (c) 2021 ${TODO: Add your comany name} and/or its affiliates.
```

File: pkg/networkservice/common/mechanisms/vxlan/vni/server.go [hunk 7]
```diff
@@ -0,0 +1,95 @@
+// Copyright (c) 2020-2021 Cisco and/or its affiliates.
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/mechanisms/vxlan/vni/server.go:1]:
```suggestion
// Copyright (c) 2021 ${TODO: Add your comany name} and/or its affiliates.
```

**Review by denis-tingaikin — COMMENTED** [review]
Also, it would be nice to cover this chain element with a simple unit test. 

See at examples: https://github.com/networkservicemesh/sdk/blob/main/pkg/networkservice/common/clienturl/server_test.go

File: pkg/networkservice/common/mechanisms/vxlan/vni/client.go (hunk 4 above)
Comment (pperiyasamy) [reply pkg/networkservice/common/mechanisms/vxlan/vni/client.go:1]: This file is copied from sdk-vpp [repo](https://github.com/networkservicemesh/sdk-vpp/tree/main/pkg/networkservice/mechanisms/vxlan/vni), so retaining the copyright here.

File: pkg/networkservice/common/mechanisms/vxlan/vni/client.go (hunk 1 above)
Comment (pperiyasamy) [reply pkg/networkservice/common/mechanisms/vxlan/vni/client.go:44]: @Bolodya1997 good find! yes, the vni client is always used along with vxlan client (this chain element is more specific to each forwarder) which sets vxlan mechanism preference and then invokes vni client. 
so adding nil check now, is that fine ?

File: pkg/networkservice/common/mechanisms/vxlan/vni/server.go (hunk 2 above)
Comment (pperiyasamy) [reply pkg/networkservice/common/mechanisms/vxlan/vni/server.go]: yes! done.

File: pkg/networkservice/common/mechanisms/vxlan/vni/server.go (hunk 3 above)
Comment (pperiyasamy) [reply pkg/networkservice/common/mechanisms/vxlan/vni/server.go]: done

File: pkg/networkservice/common/mechanisms/vxlan/vni/doc.go (hunk 5 above)
Comment (pperiyasamy) [reply pkg/networkservice/common/mechanisms/vxlan/vni/doc.go:1]: same comment as before.

File: pkg/networkservice/common/mechanisms/vxlan/vni/gen.go (hunk 6 above)
Comment (pperiyasamy) [reply pkg/networkservice/common/mechanisms/vxlan/vni/gen.go:1]: same comment as before.

File: pkg/networkservice/common/mechanisms/vxlan/vni/server.go (hunk 7 above)
Comment (pperiyasamy) [reply pkg/networkservice/common/mechanisms/vxlan/vni/server.go:1]: same comment as before.

File: pkg/networkservice/common/mechanisms/vxlan/vni/client.go (hunk 1 above)
Comment (Bolodya1997) [reply pkg/networkservice/common/mechanisms/vxlan/vni/client.go:44]: Yep, looks great.

File: pkg/networkservice/common/mechanisms/vxlan/vni/server.go (hunk 2 above)
Comment (Bolodya1997) [reply pkg/networkservice/common/mechanisms/vxlan/vni/server.go]: So actually I have meant something more specific:
```go
_, loaded := v.Map.LoadOrStore(k, &k)
conn, err := next.Server(ctx).Request(ctx, request)
if err != nil {
    if !loaded {
        v.Map.Delete(k)
    }
    return nil, err
}
```
If it was the 1 request for the Connection, it wouldn't be healed or closed on timeout, so we need this `Delete()`. But for the refresh case we probably may want to keep it until the Connection will be restored with heal or closed by heal/timeout.

File: pkg/networkservice/common/mechanisms/vxlan/vni/client.go (hunk 4 above)
Comment (denis-tingaikin) [reply pkg/networkservice/common/mechanisms/vxlan/vni/client.go:1]: OK, then be free to append a copyright holder as in example https://github.com/networkservicemesh/sdk/tree/main/.ci/license#copyright-headers-for-source-code

Comment (pperiyasamy) [reply pkg/networkservice/common/mechanisms/vxlan/vni/client.go:1]: done

File: pkg/networkservice/common/mechanisms/vxlan/vni/server.go (hunk 2 above)
Comment (pperiyasamy) [reply pkg/networkservice/common/mechanisms/vxlan/vni/server.go]: done.

**Review by denis-tingaikin — APPROVED** [review]
LGTM

**Review by edwarnicke — APPROVED** [review]

File: pkg/networkservice/common/mechanisms/vxlan/vni/server.go [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+			return nil, err
+		}
+		// If its not one already in use, set it and we are good to go
+		if _, ok := v.Map.LoadOrStore(k, &k); !ok {
+			mechanism.SetVNI(k.vni)
+			break
+		}
+	}
+	conn, err := next.Server(ctx).Request(ctx, request)
+	if err != nil {
```
Comment (d-uzlov) [inline pkg/networkservice/common/mechanisms/vxlan/vni/server.go:80]: You probably want the same `!loaded` check here, as in line 60.

**Comment by d-uzlov** [comment]
I didn't manage to finish the review in time, but I think that this check is quite important, so we probably want to fix it in another PR.

File: pkg/networkservice/common/mechanisms/vxlan/vni/server.go (hunk 8 above)
Comment (pperiyasamy) [reply pkg/networkservice/common/mechanisms/vxlan/vni/server.go:80]: @d-uzlov at this point, a new vni is generated and `!loaded` alway true, isn't it ? that's why no check is provided here.

Comment (d-uzlov) [reply pkg/networkservice/common/mechanisms/vxlan/vni/server.go:80]: Right. I guess I misread this, so this comment is invalid.

**Pull request merged**